sdks/java/core/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataGrpcMultiplexer.java

```diff
@@ -87,9 +99,24 @@ public StreamObserver<BeamFnApi.Elements> getOutboundObserver() {
     return outboundObserver;
   }
 
-  private CompletableFuture<CloseableFnDataReceiver<BeamFnApi.Elements>> receiverFuture(
+  /**
+   * Returns a future that can be awaited upon for processing elements for instruction id.
+   *
+   * @param instructionId
+   * @return null if the instruction id has been poisoned and elements should be ignored.
+   */
+  @SuppressWarnings("nullness")
+  private @Nullable CompletableFuture<CloseableFnDataReceiver<BeamFnApi.Elements>> receiverFuture(
       String instructionId) {
-    return receivers.computeIfAbsent(instructionId, (unused) -> new CompletableFuture<>());
+    return receivers.computeIfAbsent(
+        instructionId,
+        (unused) -> {
+          if (poisonedInstructionIds.getIfPresent(instructionId) != null) {
+            return null;
```

Comment:
Why do we have to special case this instead of simply always returning a completable future (even if it won't be used) as before? Or is it for better error messaging?

Reply:
This null in the computeIfAbsent is to avoid memory leak for long-running pipelines processing lots of instructions ids. If we're poisoned, it seems possible we won't unregister and thus if we add the id here it will stay in the map forever.

I removed this helper since I think it is clearer if the two cases do their own handling.